### PR TITLE
Feat! [#913]: Add event handler period

### LINKED ISSUE #913: Feature Request: Time Period for Event handler
Hi!
Thanks in advance!

I think it would be quite useful and would make the platform more versatile if a timeperiod could be applied to the execution of an event handler, since there are times, at least I have found them, when it is necessary that an event handler not be executed but if that the check of the monitored service continue to be carried out.

Regards!

### PR DESCRIPTION
**Pull request by tsadpbb**
Closes #913 \
This will add the event_handler_period configuration option to both host and service objects. This does not apply to global event handlers. If the period is not specified, the default behavior is to run the event handler.

To test: \
I added the following cfg file as `test.cfg` as well as `cfg_file=/usr/local/nagios/etc/objects/test.cfg` to the main nagios config file.
```
define host {

    use                     linux-server            ; Name of host template to use
                                                    ; This host definition will inherit all variables that are defined
                                                    ; in (or inherited by) the linux-server host template definition.
    host_name               testhost1
    alias                   testhost1
    address                 0.0.0.0
    check_command           check_always_fail
    event_handler           handle_host_always_fail!Host1
    event_handler_period    24x7
}

define host {

    use                     linux-server            ; Name of host template to use
                                                    ; This host definition will inherit all variables that are defined
                                                    ; in (or inherited by) the linux-server host template definition.
    host_name               testhost2
    alias                   testhost2
    address                 0.0.0.0
    check_command           check_always_fail
    event_handler           handle_host_always_fail!Host2
    event_handler_period    none
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost1
    service_description     Test1
    check_command           check_always_fail
    notifications_enabled   0
    event_handler           handle_service_always_fail!Test1
    event_handler_enabled   1
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost1
    service_description     Test2
    check_command           check_always_fail
    notifications_enabled   0
    event_handler           handle_service_always_fail!Test2
    event_handler_enabled   0
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost1
    service_description     Test3
    check_command           check_always_fail
    notifications_enabled   0
    event_handler           handle_service_always_fail!Test3
    event_handler_period    24x7
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost2
    service_description     Test4
    check_command           check_always_fail
    notifications_enabled   0
    event_handler           handle_service_always_fail!Test4
    event_handler_period    none
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost2
    service_description     Test5
    check_command           check_always_fail
    notifications_enabled   0
    event_handler_period    24x7
}

define command {

    command_name    check_always_fail
    command_line    $USER1$/always_fail
}

define command {

    command_name    handle_service_always_fail
    command_line    $USER1$/handle_test_fail $SERVICESTATETYPE$ $ARG1$
}

define command {

    command_name    handle_host_always_fail
    command_line    $USER1$/handle_test_fail $HOSTSTATETYPE$ $ARG1$
}
```
With always_fail and handle_test_fail being as so:
always_fail:
```
#!/bin/bash
echo "CRITICAL"
exit 0
```
handle_test_fail:
```
#!/bin/bash
echo "test! $1 $2" >> /tmp/test.txt
```

Changing `exit 0` to `exit 2` in always_fail should cause the event handlers to run. You should only see Host1, Test1, and Test3 in `/tmp/test.txt`

Test submitting commands with:
```
printf "[$(date +%s)] CHANGE_SVC_EVENT_HANDLER_TIMEPERIOD;testhost2;Test4;24x7\n" > /usr/local/nagios/var/rw/nagios.cmd

printf "[$(date +%s)] CHANGE_HOST_EVENT_HANDLER_TIMEPERIOD;testhost2;24x7\n" > /usr/local/nagios/var/rw/nagios.cmd
```

### CODE REVIEW

File: contrib/convertcfg.c [hunk 1]
```diff
@@ -382,6 +382,10 @@ int main(int argc, char **argv) {
 			if(temp_ptr != NULL && strcmp(temp_ptr, ""))
 				printf("\tevent_handler\t\t%s\n", temp_ptr);
 
+			temp_ptr = my_strsep(&temp_ptr2, ";");
+			if(temp_ptr != NULL && strcmp(temp_ptr, ""))
+				printf("\tevent_handler_period\t%s\n", temp_ptr);
+
```
Comment (sawolf) [inline contrib/convertcfg.c:388]: I don't think you have to add this or the other block in this file

**Comment by sawolf** [comment]
Hi Dylan, great work on this. I'm pretty happy with the code, just remove the blocks in the nagios-v1-to-v2 converter and we'll be good to go there. I'll do some additional testing on this soon.

**Comment by sawolf** [comment]
Based on my testing so far, I think this might need some additional work.

I used the following configuration which is similar to yours:
```
[root@localhost var]# cat /usr/local/nagios/etc/objects/test.cfg 
define host {

    use                     linux-server            ; Name of host template to use
                                                    ; This host definition will inherit all variables that are defined
                                                    ; in (or inherited by) the linux-server host template definition.
    host_name               testhost1
    alias                   testhost1
    address                 0.0.0.0
    check_command           check_always_fail
    event_handler           handle_host_always_fail!24x7
    event_handler_period    24x7
}

define host {

    use                     linux-server            ; Name of host template to use
                                                    ; This host definition will inherit all variables that are defined
                                                    ; in (or inherited by) the linux-server host template definition.
    host_name               testhost2
    alias                   testhost2
    address                 0.0.0.0
    check_command           check_always_fail
    event_handler           handle_host_always_fail!none
    event_handler_period    none
}

define host {

    use                     linux-server            ; Name of host template to use
                                                    ; This host definition will inherit all variables that are defined
                                                    ; in (or inherited by) the linux-server host template definition.
    host_name               testhost3
    alias                   testhost3
    address                 0.0.0.0
    check_command           check_always_fail
    event_handler           handle_host_always_fail!no_period_defined
}
define host {

    use                     linux-server            ; Name of host template to use
                                                    ; This host definition will inherit all variables that are defined
                                                    ; in (or inherited by) the linux-server host template definition.
    host_name               testhost4
    alias                   testhost4
    address                 0.0.0.0
    check_command           check_always_fail
    event_handler           handle_host_always_fail!no_period_defined_enabled
    event_handler_enabled   1
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost1
    service_description     Test1
    check_command           check_always_fail
    notifications_enabled   0
    event_handler           handle_service_always_fail!no_period_defined_enabled
    event_handler_enabled   1
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost1
    service_description     Test2
    check_command           check_always_fail
    notifications_enabled   0
    event_handler           handle_service_always_fail!event_handler_disabled
    event_handler_enabled   0
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost1
    service_description     Test3
    check_command           check_always_fail
    notifications_enabled   0
    event_handler           handle_service_always_fail!24x7
    event_handler_period    24x7
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost2
    service_description     Test4
    check_command           check_always_fail
    notifications_enabled   0
    event_handler           handle_service_always_fail!none
    event_handler_period    none
}

define service {

    use                     local-service           ; Name of service template to use
    host_name               testhost2
    service_description     Test5
    check_command           check_always_fail
    notifications_enabled   0
    event_handler_period    24x7
}

define command {

    command_name    check_always_fail
    command_line    $USER1$/check_dummy 2 "This check always fails"
}

define command {

    command_name    handle_service_always_fail
    command_line    /bin/echo $HOSTNAME$ $SERVICEDESC$ $SERVICESTATETYPE$ $ARG1$ >> /evhlog
}

define command {

    command_name    handle_host_always_fail
    command_line    /bin/echo $HOSTNAME$ $HOSTSTATETYPE$ $ARG1$ >> /evhlog
}
```
I also ran `touch /evhlog; chmod 777 /evhlog; tail -f evhlog` before deleting retention data and restarting `nagios` in another terminal.

Here are the log entries I get:
```
[root@localhost nagioscore]# tail -f /evhlog 
/* ...snip... */
testhost1 Test1 SOFT no_period_defined_enabled
testhost2 SOFT none
testhost1 Test1 HARD no_period_defined_enabled
testhost2 SOFT none
testhost1 Test3 HARD 24x7
testhost2 SOFT none
testhost2 SOFT none
testhost2 SOFT none
testhost2 SOFT none
testhost2 SOFT none
testhost2 SOFT none
testhost2 SOFT none
testhost2 HARD none
```
Hopefully you agree that those aren't right (and don't match your description of the behavior). I could have made a mistake in the configuration though. Let me know if you can repro/fix this.

**Comment by tsadpbb** [comment]
Those errors should be fixed now

**Comment by sawolf** [comment]
Finally got around to testing this again - my old configuration worked files are now working as expected, and the external commands are also confirmed working. Thanks again Dylan!

**Pull request merged**